### DISCUSSION TITLE
Streaming examples: Only the microphone version is bidirectional streaming.  Both HTTP examples are not.

### DISCUSSION THREAD

**Discussion by JamesKiddSmyth** [deepgram/community#581] [General help] [answered]
The streaming examples are: https://github.com/deepgram/deepgram-python-sdk/blob/main/examples/streaming/

For both of the http versions, you don't get any text back until the end.  The microphone example does work like you'd expect and I assume there is something magical about
`` microphone = Microphone(dg_connection.send)
that allows the dg_connection events to fire in real-time.

This is a bit of an issue if you want things like the utterance end event in order to have real conversational flow.

**Comment by team-deepgram** [comment]
Thanks for asking your question about Deepgram! If you didn't already include it in your post, please be sure to add as much detail as possible so we can assist you efficiently, such as:
- The `request_id` if you have a question about your requests or transcription responses.
- The features you used or the full `api.deepgram.com` URL you sent your request to, including parameters.
- Any code snippets you can share.

**Comment by davidvonthenen** [comment]
The HTTP versions are just streaming examples because they are streams. Yes, there isn't an audio going back the other way... similar to using a podcast or video, but you should get text as it's happening. This is completely legit example of streaming.

Utterance events will work for the HTTP example... they just need to be enabled:
https://github.com/deepgram/deepgram-python-sdk/blob/main/examples/streaming/microphone/main.py#L64-L67

and the message subscribed to:
https://github.com/deepgram/deepgram-python-sdk/blob/main/examples/streaming/microphone/main.py#L53-L54

and the function to handle the event defined:
https://github.com/deepgram/deepgram-python-sdk/blob/main/examples/streaming/microphone/main.py#L42-L46


Not exactly sure what the question is here.

**Reply by JamesKiddSmyth** [reply]
The live options!  thanks

**Comment by JamesKiddSmyth** [comment]
Maybe my title was misleading.  And maybe some timestamps will help. For the http version you don't get any transcription back until the end.  Microphone version starts after about 1.5 secs.   Anyway, i figured out how to make it work better on my own, but these examples don't seem to do it.

(venv) jamessmyth@Jamess-MacBook-Pro streaming % python3.11 http/main.py 
1707265861.608641: started thread 
Press Enter to stop recording...

1707265862.322751: sending to datagram
1707265862.323083: sent to datagram
1707265862.323605: sending to datagram
1707265862.323769: sent to datagram
1707265862.323882: sending to datagram
1707265862.324032: sent to datagram
1707265862.324171: sending to datagram
1707265862.3243: sent to datagram
1707265862.3243892: sending to datagram
1707265862.324508: sent to datagram
1707265862.391131: sending to datagram
1707265862.391453: sent to datagram
1707265862.41138: sending to datagram
1707265862.411607: sent to datagram
1707265862.4117289: sending to datagram
1707265862.4118762: sent to datagram
1707265862.412294: sending to datagram
1707265862.412399: sent to datagram
1707265862.41245: sending to datagram
1707265862.412525: sent to datagram
1707265862.412569: sending to datagram
1707265862.4126332: sent to datagram
1707265862.41267: sending to datagram
1707265862.4127822: sent to datagram
1707265862.412837: sending to datagram
1707265862.412916: sent to datagram
1707265862.412957: sending to datagram
1707265862.4130208: sent to datagram
1707265862.472736: sending to datagram
1707265862.4729311: sent to datagram
1707265862.473057: sending to datagram
1707265862.4731781: sent to datagram
1707265862.5004532: sending to datagram
1707265862.5005949: sent to datagram
1707265862.500693: sending to datagram
1707265862.5008268: sent to datagram
1707265862.500905: sending to datagram
1707265862.501023: sent to datagram
1707265862.501525: sending to datagram
1707265862.501647: sent to datagram
1707265862.50173: sending to datagram
1707265862.5018492: sent to datagram
1707265862.5019221: sending to datagram
1707265862.502042: sent to datagram
1707265862.502111: sending to datagram
1707265862.502217: sent to datagram
1707265862.5022788: sending to datagram
1707265862.502361: sent to datagram
1707265862.502423: sending to datagram
1707265862.502505: sent to datagram
1707265862.5025642: sending to datagram
1707265862.502669: sent to datagram
1707265862.549841: sending to datagram
1707265862.55008: sent to datagram
1707265862.6321018: sending to datagram
1707265862.6322901: sent to datagram
1707265862.705656: sending to datagram
1707265862.705838: sent to datagram
1707265862.7742138: sending to datagram
1707265862.774465: sent to datagram

1707265864.779826: thread joined
1707265866.5719109: message received: israel has confirmed that thirty one of the remaining hostage
1707265866.649606: dg_connection finish 
Finished
(venv) jamessmyth@Jamess-MacBook-Pro streaming % python3.11 microphone/main.py 
1707265900.7835898: mic started
Press Enter to stop recording...

1707265901.382825: sending audio to datagram
1707265901.384199: sent  audio to datagram
1707265901.456983: speech started
1707265901.894651: sending audio to datagram
1707265901.8957598: sent  audio to datagram
1707265901.996781: message speaker: Israel
1707265902.4067378: sending audio to datagram
1707265902.4074612: sent  audio to datagram
1707265902.918735: sending audio to datagram
1707265902.9195678: sent  audio to datagram
1707265903.021908: message speaker: Israel has confirmed that
1707265903.430797: sending audio to datagram
1707265903.4316428: sent  audio to datagram
1707265903.9424489: sending audio to datagram
1707265903.9432771: sent  audio to datagram
1707265904.1029608: message speaker: Israel has confirmed that thirty one of the
1707265904.454711: sending audio to datagram
1707265904.455591: sent  audio to datagram

1707265904.918082: mic finishing
1707265905.07669: mic finished
1707265905.2404668: message speaker: Israel has confirmed that thirty one of the remaining high


MetadataResponse(type='Metadata', transaction_key='deprecated', request_id='ab490ecf-e442-4e52-a0b5-ab60934737a0', sha256='d9d2e95ea9e08b991467d2622dc738b201388272c92fc1803dab8c908df3b835', created='2024-02-07T00:31:39.917Z', duration=3.584875, channels=1, models=['1dbdfb4d-85b2-4659-9831-16b3c76229aa'], model_info={'1dbdfb4d-85b2-4659-9831-16b3c76229aa': ModelInfo(name='2-general-nova', version='2024-01-11.36317', arch='nova-2')}, extra=None)

**Comment by davidvonthenen** [comment]
yea.. that shouldnt be the case. when I run the example, you should see the words getting printed as they happen:

```
(deepgram-python-sdk) vonthd@vonthds-MacBook-Pro-2:http$ python main.py 
Press Enter to stop recording...

speaker: and shubundu anuzzo and thank you for your ears
speaker: if you want to be in touch with the show you can email the artseller at b b c
speaker: dot c o dot u k we'd love to hear from you
speaker: for now
speaker: from me nikki brady studio manager duncan hannant and producer oliver jones
speaker: see you next week
speaker: this is the bbc world service
speaker: in conversation with acclaimed italian physicist and writer

speaker: carlo rovelli i fell in love with science in my late twenties
Finished
(deepgram-python-sdk) vonthd@vonthds-MacBook-Pro-2:http$ 
```

**Reply by JamesKiddSmyth** [reply]
but those are full transcription, not streamed back.

**Reply by davidvonthenen** [reply]
Those are happening in realtime (aka streamed) on the websockets streaming the BBC live stream.

**Reply by davidvonthenen** [reply, marked as answer]
@JamesKiddSmyth  If you run the example and also listen to the broadcast, you can see that it's happening in real time. Let me know if you still need help.

**Reply by davidvonthenen** [reply]
Reach out in Discord if you are still having problems.

**Comment by davidvonthenen** [comment]
You can also see the transcription happening in your logs as well:

```
1707265901.382825: sending audio to datagram
1707265901.384199: sent audio to datagram
1707265901.456983: speech started
1707265901.894651: sending audio to datagram
1707265901.8957598: sent audio to datagram
>>>>>>>>> 1707265901.996781: message speaker: Israel <<<<<<<<<<<
1707265902.4067378: sending audio to datagram
1707265902.4074612: sent audio to datagram
1707265902.918735: sending audio to datagram
1707265902.9195678: sent audio to datagram
>>>>>>>>> 1707265903.021908: message speaker: Israel has confirmed that <<<<<<<<<<<
1707265903.430797: sending audio to datagram
1707265903.4316428: sent audio to datagram
1707265903.9424489: sending audio to datagram
1707265903.9432771: sent audio to datagram
>>>>>>>>> 1707265904.1029608: message speaker: Israel has confirmed that thirty one of the  <<<<<<<<<<<
1707265904.454711: sending audio to datagram
1707265904.455591: sent audio to datagram

1707265904.918082: mic finishing
1707265905.07669: mic finished
1707265905.2404668: message speaker: Israel has confirmed that thirty one of the remaining high

MetadataResponse(type='Metadata', transaction_key='deprecated', request_id='ab490ecf-e442-4e52-a0b5-ab60934737a0', sha256='d9d2e95ea9e08b991467d2622dc738b201388272c92fc1803dab8c908df3b835', created='2024-02-07T00:31:39.917Z', duration=3.584875, channels=1, models=['1dbdfb4d-85b2-4659-9831-16b3c76229aa'], model_info={'1dbdfb4d-85b2-4659-9831-16b3c76229aa': ModelInfo(name='2-general-nova', version='2024-01-11.36317', arch='nova-2')}, extra=None)
```

If you want to skip the final message, aka `Israel has confirmed that thirty one of the remaining high`, you need to turn on `interim_results` like this just like in the microphone example:
https://github.com/deepgram/deepgram-python-sdk/blob/main/examples/streaming/microphone/main.py#L65

Documentation for `interim_results`:
https://developers.deepgram.com/docs/interim-results

**Reply by JamesKiddSmyth** [reply]
for the microphone example, not the http example.

**Reply by davidvonthenen** [reply]
My bad... but the HTTP example above is in realtime
